### PR TITLE
#165412931 Add search functionalities using filters

### PR DESCRIPTION
**Pull request by djallas**
#### What does this PR do?
Fetch articles by filters

#### Description of Task to be completed?
**Allowed filters**

1. author
`?author=john`

2. keyword
`?keyword=lorum`

3. tags
`?tag=news`

**Note that:**
- default limit is **20**
- default offset is **0** 

#### How should this be manually tested?
After cloning the repository, switch into it and run  `npm test` in the terminal 

**Endpoints**
/GET `/api/v1/articles`

**Options**

1. by tag: `?tag=news`
2. by author: `?author=doe`
3. by keyword: `?keyword=ipsum`
4. by keyword and author: `?keyword=lorum&author=doe`
5. by keyword and tag: `?keyword=lorum&tag=news`
6. by author and tag: `?author=doe&tag=news`
7. by author, keyword and tag: `?author=doe&keyword=ipsum&tag=news`

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#165412931](https://www.pivotaltracker.com/story/show/165412931)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/ArticleController.js [hunk 1]
```diff
@@ -35,15 +35,27 @@ export default class ArticleController {
...
     const queries = {
       page: parseInt(limit, 0) || 20, // if pagination is not speficied, fetch 20 recent articles
-      offset: offset || 0 // default offset value: 0
+      offset: offset || 0, // default offset value: 0
+      keyword, // search keyword
+      author, // article author
+      tag // tags of the article
     };
-    const allArticle = await Article.getAll(queries.page, queries.offset);
+    const allArticle = await Article.getAll(
```
[the remark this answers is not in the document]
Comment (djallas) [reply src/controllers/ArticleController.js]: @e-liyai I applied the change, you can review!

File: src/queries/articles/getAll.js [hunk 2]
```diff
@@ -1,21 +1,31 @@
...
  * @param {object} offset offset for query
+ * @param {object} keyword search keyword
+ * @param {object} author search by author
+ * @param {object} tag search by tag
  * @param {object} condition condition for query
  * @returns {object} Object representing the response returned
  */
-export default async (limit, offset) => {
-  const condition = { status: { [Op.ne]: 'deleted' } };
+export default async (limit, offset, keyword, author, tag) => {
```
Comment (e-liyai) [inline src/queries/articles/getAll.js]: we need to rethink this logic, too many if statements

File: src/controllers/AuthLocalController.js [hunk 3]
```diff
@@ -18,27 +18,23 @@ export default class AuthLocalController {
...
-      if (isUser) {
-        return res.status(status.EXIST).send({ error: 'Sorry, this account already exists' });
-      }
-      const newUser = await User.create({
-        firstName,
-        lastName,
-        username,
-        email,
-        password: helper.password.hash(req.body.password)
+    const isUser = await helper.isUser({ email });
```
Comment (e-liyai) [inline src/controllers/AuthLocalController.js]: why did you remove the try catch block

File: src/helpers/parameters/duplication.js [hunk 4]
```diff
@@ -4,13 +4,24 @@
  * @returns {array} an array of errors or an empty array if no error
```
Comment (e-liyai) [inline src/helpers/parameters/duplication.js]: why is this file named duplication?

File: src/controllers/AuthLocalController.js (hunk 3 above)
Comment (djallas) [reply src/controllers/AuthLocalController.js]: Hello @e-liyai, this file was changed after **rebasing** from **develop**. There were errors in tests, I asked help from my colleague @sengayire who worked on **Authentication** and **User management** to fix this for me. Now, errors are being caught from middlewares and helpers.

File: src/helpers/parameters/duplication.js (hunk 4 above)
Comment (djallas) [reply src/helpers/parameters/duplication.js]: @e-liyai I renamed the file to `checkDuplication.js`

File: src/tests/routes/userArticles.test.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  describe('ARTICLES', () => {
+    let createdArticle2 = '';
+    before(async () => {
+      // create article 1
+      delete article1.id;
+      delete article1.readTime;
+      article1.slug = 'baller-wayde-hasli23f1';
+      article1.userId = createdUser.id;
+      await db.Article.create(article1, { logging: false });
+      // console.log('rw1', createdArticle1);
```
Comment (e-liyai) [inline src/tests/routes/userArticles.test.js]: remove comments

File: src/tests/routes/userArticles.test.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import chai from 'chai';
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (e-liyai) [inline src/tests/routes/userArticles.test.js]: if this check is no longer applicable remove

File: src/helpers/parameters/checkDuplication.js [hunk 7]
```diff
@@ -4,13 +4,24 @@
...
+    limit, offset, keyword, author, tag
+  } = query;
   const errors = {};
   if (Array.isArray(limit)) {
     errors.limit = 'Limit can not be declared more than once';
   }
   if (Array.isArray(offset)) {
     errors.offset = 'Offset can not be declared more than once';
   }
+  if (Array.isArray(keyword)) {
```
Comment (e-liyai) [inline src/helpers/parameters/checkDuplication.js]: I believe all these checks can be placed in one or statement and return a message that encapsulates all instances

File: src/helpers/parameters/checkDuplication.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+/**
```
Comment (e-liyai) [inline src/helpers/parameters/checkDuplication.js]: where are you using this check

File: src/queries/articles/getUserArticles.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import db from '../../models';
+// import * as filters from '../../helpers/searchArticleFilters';
```
Comment (e-liyai) [inline src/queries/articles/getUserArticles.js]: if no longer in use, delete this

File: src/routes/auth/local.js [hunk 10]
```diff
@@ -2,11 +2,11 @@ import express from 'express';
 import AuthLocalController from '../../controllers/AuthLocalController';
 import asyncHandler from '../../middlewares/asyncHandler';
 import verifyToken from '../../middlewares/verifyToken';
-// import validateUser from '../../middlewares/validateUser';
+import validateUser from '../../middlewares/validateUser';
 
 const router = express.Router();
 
-router.post('/signup', AuthLocalController.signup);
+router.post('/signup', validateUser, AuthLocalController.signup);
```
Comment (e-liyai) [inline src/routes/auth/local.js]: how are you error handling this post?

File: src/helpers/parameters/checkDuplication.js (hunk 8 above)
Comment (djallas) [reply src/helpers/parameters/checkDuplication.js]: In route: `line 20`
https://github.com/andela/ninjas-ah-backend/blob/develop/src/routes/api/articles.js

File: src/routes/auth/local.js (hunk 10 above)
Comment (sengayire) [reply src/routes/auth/local.js]: Hello @e-liyai. I worked on signing up a new user. 
Previously in `develop`,  we did not have an error handler. For now, I introduced an error handling algorithm in #34 which is not yet merged
